include: zephyr: dt-bindings: clock: factorize STM32_* clock macros

etienne-lms opened the pull request:
Factorize definitions of the macros used to describe in the STM32 boards device tree the RCC register dedicated to source clock selection configuration, instead of replicating the macros per SoC header file.

etienne-lms committed:
include: zephyr: dt-bindings: clock: factorize STM32_* clock source macros

Factorize definitions of the macros used to describe in the STM32
boards device tree the RCC register dedicated to clock source selection
configuration, instead of replicating the macros per SoC header file.
For that purpose, use DT cell bit fields that matches all SoCs.

This change also factorizes STM32_MCO_CFGR_* macros that use the same
packing layout for defining selection of the MCO clocks and for MCO
pre-scaling factor on SoCs that support the feature.

By the way, reorder argument description in macro inline description
comment in a standard way (first argument first).

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
include: zephyr: dt-bindings: clock: fix missing STM32 header files

Add inclusion of stm32f4_clock.h header file that is missing
in stm32f410 and stm32f427 DT bindings clock header files.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

gautierg-st commented on include/zephyr/dt-bindings/clock/stm32_common_clocks.h:
```diff
@@ -20,35 +20,58 @@
+#define STM32_CLOCK_MASK_SHIFT		STM32_DT_RCC_MASK_SHIFT
+#define STM32_CLOCK_VAL_MASK		STM32_DT_RCC_VAL_MASK
+#define STM32_CLOCK_VAL_SHIFT		STM32_DT_RCC_VAL_SHIFT
+#define STM32_DOMAIN_CLOCK(val, mask, shift, reg) \
+	STM32_DT_RCC_CLOCK_SELECT((val), (mask), (shift), (reg))
```
Likewise, is there a point to still distinguish `STM32_DOMAIN_CLOCK` and `STM32_MCO_CFGR`?
Since we're simplifying and factorizing, I'd go the full route and use only the new `STM32_DT_RCC_CLOCK_SELECT` in every file.
@erwango What do you think?

etienne-lms replied:
I would tend to agree with you.

erwango replied:
Fine with me.
Let's update the full range of stm32*_clocks.h files then.

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32_common_clocks.h:
```diff
@@ -20,35 +20,58 @@
+	 (((val) & STM32_DT_RCC_VAL_MASK) << STM32_DT_RCC_VAL_SHIFT))
+
+/* STM32_CLOCK_* macros, defined for convenience */
+#define STM32_CLOCK_REG_MASK		STM32_DT_RCC_REG_MASK
+#define STM32_CLOCK_REG_SHIFT		STM32_DT_RCC_REG_SHIFT
+#define STM32_CLOCK_SHIFT_MASK		STM32_DT_RCC_SHIFT_MASK
+#define STM32_CLOCK_SHIFT_SHIFT		STM32_DT_RCC_SHIFT_SHIFT
+#define STM32_CLOCK_MASK_MASK		STM32_DT_RCC_MASK_MASK
+#define STM32_CLOCK_MASK_SHIFT		STM32_DT_RCC_MASK_SHIFT
+#define STM32_CLOCK_VAL_MASK		STM32_DT_RCC_VAL_MASK
+#define STM32_CLOCK_VAL_SHIFT		STM32_DT_RCC_VAL_SHIFT
```
`STM32_CLOCK_xxx_GET()` macros, defined in *include/.../stm32_clock_control.h*, still rely on these macros. That said, it could be updated to use the new `STM32_DT_RCC_xxx_[MASK|SHIFT]` macros.

etienne-lms committed:
include: zephyr: dt-bindings: clock remove unused STM32 macros

Remove now unused macros STM32_DOMAIN_CLOCK(), STM32_CLOCK_*_SHIFT,
STM32_CLOCK_*_MASK, STM32_MCO_CFGR(), STM32_MCO_CFGR_*_SHIFT and
STM32_MCO_CFGR_REG_MASK macros, replaced with STM32_DT_CLOCK_SELECT(),
STM32_DT_CLKSEL_*_SHIFT and STM32_DT_CLKSEL_*_MASK macros.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms committed:
drivers: clock: stm32: factorize clock selection macros

Factorize STM32_CLOCK_*_GET() and STM32_MCO_CFGR_*_GET() macros
into a single series of STM32_DT_CLKSEL_*_GET() macros based on
recently introduced new common macros STM32_DT_CLKSEL_*_SHIFT and
STM32_DT_CLKSEL_*_MASK.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>

etienne-lms commented:
Series updated as per https://github.com/zephyrproject-rtos/zephyr/pull/84829#discussion_r1940765409 discussion outcome.
I've changed the macro names to `STM32_DT_RCC_CLKSRC_xxx` to clarify they are dedicated to DT binding data related to clock source selection applicable the STM32 RCC SoC registers.

erwango commented on include/zephyr/dt-bindings/clock/stm32_common_clocks.h:
```diff
@@ -20,35 +20,34 @@
-#define STM32_MCO_CFGR_VAL_MASK     0x1FU
-#define STM32_MCO_CFGR_VAL_SHIFT    27U
+/** Helper macros to pack RCC clock source selection register info in the DT */
+#define STM32_DT_RCC_CLKSRC_REG_MASK	0xFFFFU
+#define STM32_DT_RCC_CLKSRC_REG_SHIFT	0U
+#define STM32_DT_RCC_CLKSRC_SHIFT_MASK	0x3FU
+#define STM32_DT_RCC_CLKSRC_SHIFT_SHIFT	16U
+#define STM32_DT_RCC_CLKSRC_MASK_MASK	0x1FU
+#define STM32_DT_RCC_CLKSRC_MASK_SHIFT	22U
+#define STM32_DT_RCC_CLKSRC_VAL_MASK	0x1FU
+#define STM32_DT_RCC_CLKSRC_VAL_SHIFT	27U
```
Since the scope of these macros is now quite large (MCO and others), it would make sense to make the name more generic.
So replace `CLKSRC` by `CLK` or other generic proposal (could be `REG` but then we'd have `REG_REG` occurences).

etienne-lms replied:
I can replace `RCC_CLKSRC` with `CLKSEL` since these are expected to be used for clock selection configuration: `STM32_DT_CLKSEL_xxx`.

erwango replied:
> I can replace `RCC_CLKSRC` with `CLKSEL` since these are expected to be used for clock selection configuration: `STM32_DT_CLKSEL_xxx`.

Do you propose this also applies to `STM32_DT_RCC_CLKSRC_SELECT` also ?

erwango reviewed (changes requested):
Otherwise LGTM

etienne-lms commented on include/zephyr/dt-bindings/clock/stm32f410_clock.h:
```diff
@@ -6,6 +6,8 @@
 #ifndef ZEPHYR_INCLUDE_DT_BINDINGS_CLOCK_STM32F410_CLOCK_H_
 #define ZEPHYR_INCLUDE_DT_BINDINGS_CLOCK_STM32F410_CLOCK_H_
 
+#include "stm32_common_clocks.h"
```
This change was rather for consistency. I get the point.
That said, if *stm32f410_clock.h* and *stm32f427_clock.h* are expected to be included after *stm32f4_clock.h* (as mentioned at line 31), I think here I should replace line 9 with `#include "stm32f4_clock.h"`.

erwango replied:
`#include "stm32f4_clock.h"` is already included in `stm32f4.dtsi`

etienne-lms replied:
I mean *stm32f410_clock.h* and *stm32f427_clock.h* should include *stm32f4_clock.h*, wherever they are included from. Less fragile IMHO.

etienne-lms commented:
Typo fixed in *drivers/clock_control/clock_stm32_mco.c*.
Previous twister test campaign failed on "*mini_stm32h743 sample.video.capture_to_lvgl.shield.weact_ov2640_cam_module*": I could not reproduce the issue. I'll need to check that again if it still fails.

kartben commented:
@etienne-lms @gautierg-st 
needs rebase - thanks!

etienne-lms commented:
Indeed. Thanks for the notif.

etienne-lms committed:
include: zephyr: dt-bindings: clock: use STM32 common clock source macros

Replace STM32_DOMAIN_CLOCK() and STM32_MCO_CFGR() macros in
stm32 DT clock bindings header file with the common macro
STM32_DT_CLOCK_SELECT().

By the way, add parentheses on argument where it mas missing.

Signed-off-by: Etienne Carriere <etienne.carriere@st.com>
